src/dxgi.rs

```diff
@@ -39,13 +42,53 @@ pub enum AlphaMode {
     ForceDword = dxgi1_2::DXGI_ALPHA_MODE_FORCE_DWORD,
 }
 
+pub type InfoQueue = WeakPtr<dxgidebug::IDXGIInfoQueue>;
+
 pub type Adapter1 = WeakPtr<dxgi::IDXGIAdapter1>;
+pub type Adapter2 = WeakPtr<dxgi1_2::IDXGIAdapter2>;
+pub type Adapter3 = WeakPtr<dxgi1_4::IDXGIAdapter3>;
+pub type Adapter4 = WeakPtr<dxgi1_6::IDXGIAdapter4>;
+crate::weak_com_inheritance_chain! {
+    #[derive(Debug, Copy, Clone, PartialEq, Hash)]
+    pub enum DxgiAdapter {
+        Adapter1(dxgi::IDXGIAdapter1), from_adapter1, as_adapter1, adapter1;
+        Adapter2(dxgi1_2::IDXGIAdapter2), from_adapter2, as_adapter2, unwrap_adapter2;
```

Comment:
it seems quite silly to pass in all the boilerplate like `as_adapterN` and `unwrap_adapterN`. Can't those instead be implementations of some type-generic traits? I.e. `impl ComAs<dxgi1_2::IDXGIAdapter2> for Self {...}`. This would make all those method names unnecessary, I hope.

Reply:
I did this to prevent the need to importing traits at all levels, but if that is okay, then we likely could. Though at all callsites we'd need type annotations as inference wouldn't be able to figure out which one to use, which might be too annoying to use.

Reply:
For imports, we'd only need to import 2 traits, so seems ok (considering how much stuff we already import).
For usage, yes, it could be too verbose. Are you confident that this would be the case?

Reply:
Yeah I'm fairly certain:

```rust
if let Some(factory5) = factory.as_com() {
   ... 
} 
```

There's no type to cause inference to work right here.

Reply:
yes, I understand it will be more verbose. I'm asking if you are sure that this verbosity will be too much of us, or maybe it's ok?

Reply:
Ah, I see what you're asking. I think the verbosity would be a little much for us, it would require doing `AsCom::<dxgi1_5::Factory5>::as_com(&factory)` to get type annotations in there. I think the small amount of boilerplate at the declaration site is worth the farely large boilerplate savings throughout.

Reply:
ok, I'll be on a lookout for this :)